marimo/_runtime/executor.py

```diff
@@ -103,39 +150,54 @@ async def execute_cell_async(
         pass
 
 
+# NOTE: Prefer to inherit from DefaultExecutor, or call the static execute
+# methods directly since DefaultExecutor utilizes so unorthodox code formmating
+# for prettier stack traces.
 @register_execution_type("relaxed")
 class DefaultExecutor(Executor):
     @staticmethod
     async def execute_cell_async(
         cell: CellImpl,
         glbls: dict[str, Any],
-        _graph: Optional[DirectedGraph] = None,
+        graph: Optional[DirectedGraph] = None,
     ) -> Any:
         if cell.body is None:
             return None
         assert cell.last_expr is not None
-
-        if _is_coroutine(cell.body):
-            await eval(cell.body, glbls)
-        else:
-            exec(cell.body, glbls)
-
-        if _is_coroutine(cell.last_expr):
-            return await eval(cell.last_expr, glbls)
-        else:
-            return eval(cell.last_expr, glbls)
+        try:
+            if _is_coroutine(cell.body):
+                await eval(cell.body, glbls)
+            else:
+                exec(cell.body, glbls)
+
+            if _is_coroutine(cell.last_expr):
+                return await eval(cell.last_expr, glbls)
+            else:
+                return eval(cell.last_expr, glbls)
+        except NameError as e:
+            raise_name_error(graph, e)
+        except (BaseException, Exception) as e:
+            # Raising from a BaseException will fold in the stactrace prior
+            # to execution
+            raise MarimoBaseException from e
```

Comment:
The discussion thread posted doesn't make a great case for manipulating the stacktrace as a practice- but the examples posted are pertinent

The **Jinja** example the trace is changed to make it look as if it comes from the template file.

In **IPython/ jupyter** the raw trace is not carried but the relevant stack info and formatted into their own stacktrace format (also useful for different kernel types):

![image](https://github.com/user-attachments/assets/8a762e1a-95a1-40f9-9c97-b353e1fc7fbf)

This is also needed a bit more since Jupyter can point to code that no longers exists:

![image](https://github.com/user-attachments/assets/6249578d-f346-47c2-bfe9-2882b9d99bcf)

Stack traces from files can also give a bit more context:

![image](https://github.com/user-attachments/assets/9e44efc6-5a35-4e66-b752-d7d19cf8e05d)

I believe the relevant code is the aptly named [`ultratb`](https://github.com/ipython/ipython/blob/main/IPython/core/ultratb.py).

---

ideally, I think marimo still wants to provide python like stack traces since it's actually python, but maybe it can have a more managed approach? I see this as a follow up PR done in FE, but how's this look for a goal with links? Or is the ultratrace closer to what we'd want? Is the formatting consistent in script mode?

![image](https://github.com/user-attachments/assets/7d7a60f6-ade5-49e4-8154-5c18915d5bf6)

If the trace is going to be totally manipulated front-end like jupyter, I do think the complexity +  this is a compelling reason not to use MarimoBaseException/ MarimoRuntimeException

I'm good to revert this, but just on a thought regarding this comment:

The error handling could be moved to run

```python
def run():
    try:
       # ...
    except MarimoRuntimeError as e:
        raise e.__cause__
    except Exception as e:
        # Known for certain to come from marimo itself
        raise MarimoScriptException("Congrats! You definitely found a marimo bug. Please report") from e
```

Which is still useful if an app is used from another script, and raises the raw error unless something is actually wrong

Reply:
@dmadisetti , thanks for sharing the context and the mock-up for the future design. The screenshot LGTM; I think  the user doesn't need to see the first frame (`executor.py`), but it's okay if they do.

`run()` logic with error handling looks good. 

No need to revert! If the error handling for running as as script is updated as per your example, that's good with me. I just didn't want to push more complexity onto the user.